### PR TITLE
#42595 Form flow expressions

### PR DESCRIPTION
**Pull request by Klaas-Ritense**

### CODE REVIEW

File: release-notes/major9/09.26.0/valtimo-backend-libraries.md [hunk 1]
```diff
@@ -10,10 +10,10 @@ The following features were added:
   Additionally, case settings can be loaded from a file when the corresponding document definition is created. For more
   information on how to create a case definition when a document definition is created, see [here](/using-valtimo/case/creating-case-settings.md).
 
-* **New feature2**
+* **Form Flow bean for completing the task**
```
Comment (marijnritense) [inline release-notes/major9/09.26.0/valtimo-backend-libraries.md:13]: Maybe rephrase to: Form flow bean for completing an associated task.

**Review by marijnritense — DISMISSED** [review]

File: reference/modules/form-flow.md [hunk 2]
```diff
@@ -32,4 +32,51 @@ The following properties are available in a SpEL context:
 | `additionalProperties` | -                | `Map<String, Any>`       | Additional properties stored as part of the form flow instance. |
 
 
+## SpEL bean whitelisting
+SpEL by default can call any kind of bean to offer extra functionality to expressions. You can for instance call a
+service to send an email, create a document based on a template etcetera. This is a powerful feature, but you can't
+control what kind of beans might be called. Together with the possibility to use / fill expressions based user input,
+and we have a major security concern.
```
Comment (theo-ritense) [inline reference/modules/form-flow.md:39]:
```suggestion
there is a major security concern.
```

File: reference/modules/form-flow.md [hunk 3]
```diff
@@ -32,4 +32,51 @@ The following properties are available in a SpEL context:
...
+property.
+
+If for example an evaluation calls a service, but that service is not whitelisted, SpEL will not be called to evaluate
+and parse the expression. Valtimo will filter out any expressions that could be a security risk.
+
+Non-bean calls are still allowed. You can for example just use a string or a math equation within SpEL as well, these do
+not require a whitelisting. Examples of what you can enter as an expression in SpEL can be found in the spring
+documentation (https://spring.getdocs.org/en-US/spring-framework-docs/docs/spring-core/expressions/expressions-language-ref.html).
+
+All expression on this page are displayed in code within the function .parseExpression(). Anything you find within this
```
Comment (theo-ritense) [inline reference/modules/form-flow.md:53]:
```suggestion
All expressions passed as an argument to `parseExpression(...)` on this page can also be set in the form flow definition between `${}`
```

File: reference/modules/form-flow.md [hunk 4]
```diff
@@ -32,4 +32,51 @@ The following properties are available in a SpEL context:
...
+function can be set in the form flow definition with ${ }.
```
Comment (theo-ritense) [inline reference/modules/form-flow.md:54]:
```suggestion
```

File: reference/modules/form-flow.md [hunk 5]
```diff
@@ -32,4 +32,50 @@ The following properties are available in a SpEL context:
...
+```spel
+${valtimoFormFlow.completeTask(additionalProperties, step.submissionData)}
+```
+
+The method below completes a Camunda user task and saves the submission data into a location of your choosing. The third
+parameter decides where the submission data should be saved. The third parameter expects a map from which the key
+defines the location where the data should be saved, and the value is a location inside the `step.submissionData`.
+
+```spel
+${valtimoFormFlow.completeTask(additionalProperties, step.submissionData, {'doc:/address/streetName':'/street', 'pv:approved':'/approval'})}
```
Comment (ivo-ritense) [inline reference/modules/form-flow.md:78]: I think it might be good to have a example submission and document before and after to go with this example to make it more clear.

Comment (ivo-ritense) [reply reference/modules/form-flow.md:78]: Some questions that i have that the user might also have so they might need to be documented as well:
- When the destination path does not exist, will it also make all objects in between, or will you get an error?
- Can you copy over a whole object in 1 go, or do you have to go property by property?
- Does this support arrays as well?
-

File: using-valtimo/form-flow/create-form-flow-definition.md [hunk 6]
```diff
@@ -203,7 +203,7 @@ available can be found [here](/reference/modules/form-flow.md#available-properti
                     "definition": "end-form"
                 }
             },
-            "onComplete": ["${someDocumentService.saveFormFlowResult(additionalProperties, step.submissionData)}"]
+            "onComplete": ["valtimoFormFlow.completeTask(additionalProperties, step.submissionData, {'doc:/':''})"]
```
Comment (ivo-ritense) [inline using-valtimo/form-flow/create-form-flow-definition.md:206]: is this change correct? it does not match the code in the test implementation.

Comment (Klaas-Ritense) [reply using-valtimo/form-flow/create-form-flow-definition.md]: It is correct. But I removed it, since it doesn't make the example more clear.

File: reference/modules/form-flow.md (hunk 5 above)
Comment (Klaas-Ritense) [reply reference/modules/form-flow.md:103]: Added

**Review by ivo-ritense — APPROVED** [review]
